### PR TITLE
Fix Basic Station Issues

### LINKED ISSUE #796: Various fixes for Basic Station LNS
#### Summary

This is a collection of issues that I found while testing LNS on metal (TTIG).

- [x] A couple of simple JSON format mismatches
- [x] `SX1301_conf` adaptation
- [x] DL timing fix
- [x] Misc fixes

#### Steps to Reproduce

1. Connect TTIG to server.
2. Check gateway logs.

#### What do you see now?

Each error has a different effect on the Gateway.  I'll add comments in the code to non-obvious fixes.

#### What do you want to see instead?

Gateway works.

#### Environment

Basic Station LNS on bare-metal, `v3.0.3`.
#### How do you propose to implement this?

 Already fixed. This issue is to document them.

#### Can you do this yourself and submit a Pull Request?

Yes a follow-up PR is on the way.

### PR DESCRIPTION
**Pull request by KrishnaIyer**
#### Summary

Closes #796 

#### Changes

- Fixes reported issues.
~- Adds custom values in `io.Connection`.~
- Fix tests.

#### Notes for Reviewers

- After thinking about #794, that won't work since the MSB of `XTime` is in the opposite direction of downlink tokens.  It seems to me that this belongs in the connection per gateway so added a custom value there.

### CODE REVIEW

File: pkg/gatewayserver/io/basicstationlns/messages/upstream_test.go [hunk 1]
```diff
@@ -101,7 +101,7 @@ func TestMarshalJSON(t *testing.T) {
 				t.Fatalf("Unexpected error: %v", err)
 			}
 			if !a.So(msg, should.Resemble, tc.Expected) {
-				t.Fatalf("Unexpected message: %v", msg)
+				t.Fatalf("Unexpected message: %v\n, Expected: %s", string(msg), string(tc.Expected))
```
Comment (johanstokking) [inline pkg/gatewayserver/io/basicstationlns/messages/upstream_test.go]: `t` has enough context (i.e. test name) to know `tc.Expected`

File: pkg/gatewayserver/io/io.go [hunk 2]
```diff
@@ -64,7 +64,8 @@ type Connection struct {
 	lastStatusTime,
 	lastUplinkTime,
 	lastDownlinkTime int64
-	lastStatus atomic.Value
+	lastStatus,
+	custom atomic.Value // custom value used by each front end.
```
Comment (johanstokking) [inline pkg/gatewayserver/io/io.go]: Let's not do this

File: pkg/gatewayserver/io/io_test.go [hunk 3]
```diff
@@ -371,4 +372,20 @@ func Test(t *testing.T) {
 			a.So(time.Since(last), should.BeLessThan, timeout)
 		})
 	}
+
+	for i, tc := range []struct {
```
Comment (johanstokking) [inline pkg/gatewayserver/io/io_test.go]: We don't need a test table for one entry. Also if the entry has one field, there's no need for a struct.

File: pkg/gatewayserver/io/basicstationlns/messages/messages_test.go [hunk 4]
```diff
@@ -190,22 +242,73 @@ func TestGetRouterConfig(t *testing.T) {
...
 				if tc.ErrorAssertion == nil || !a.So(tc.ErrorAssertion(err), should.BeTrue) {
 					t.Fatalf("Unexpected error: %v", err)
 				}
 			} else if tc.ErrorAssertion != nil {
 				t.Fatalf("Expected error")
 			} else {
+				cfg.MuxTime = 0
 				if !a.So(cfg, should.Resemble, tc.Cfg) {
-					t.Fatalf("Invalid config: %v", cfg)
+					t.Fatalf("Invalid config: %v\n", cfg)
```
Comment (johanstokking) [inline pkg/gatewayserver/io/basicstationlns/messages/messages_test.go]: Why?

File: pkg/gatewayserver/io/basicstationlns/messages/downstream.go [hunk 5]
```diff
@@ -83,9 +84,10 @@ func FromDownlinkMessage(ids ttnpb.GatewayIdentifiers, down ttnpb.DownlinkMessag
 
 	dnmsg.DeviceClass = uint(ttnpb.CLASS_A)
 	// Estimate the xtime based on the timestamp; xtime = timestamp - (rxdelay+1).
+
 	t := time.Unix(int64(scheduledMsg.Timestamp), 0)
-	offset := time.Duration((dnmsg.RxDelay + 1)) * time.Second
-	dnmsg.XTime = t.Add(-offset).Unix()
+	offset := time.Duration((dnmsg.RxDelay)) * time.Second * (1e6)
+	dnmsg.XTime = xTimeUpper&0x7FFFFFFF00000000 | (t.Add(-offset).Unix())
```
Comment (johanstokking) [inline pkg/gatewayserver/io/basicstationlns/messages/downstream.go]: No parentheses needed for the second part

Comment (johanstokking) [reply pkg/gatewayserver/io/basicstationlns/messages/downstream.go]: So these are seconds?

File: pkg/gatewayserver/io/basicstationlns/messages/downstream.go [hunk 6]
```diff
@@ -56,20 +54,23 @@ func (dnmsg DownlinkMessage) MarshalJSON() ([]byte, error) {
 }
 
 // FromDownlinkMessage translates the ttnpb.DownlinkMessage to LNS DownlinkMessage "dnmsg".
-func FromDownlinkMessage(ids ttnpb.GatewayIdentifiers, down ttnpb.DownlinkMessage, dlToken int64, dlTime time.Time) DownlinkMessage {
+func FromDownlinkMessage(ids ttnpb.GatewayIdentifiers, down ttnpb.DownlinkMessage, dlToken int64, dlTime time.Time, xTimeUpper int64) DownlinkMessage {
```
Comment (johanstokking) [inline pkg/gatewayserver/io/basicstationlns/messages/downstream.go]: Isn't this `xtime`?

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 7]
```diff
@@ -203,21 +204,27 @@ func (s *srv) handleTraffic(c echo.Context) error {
...
+				}
+				dnmsg := messages.FromDownlinkMessage(ids, *down, int64(s.tokens.Next(down.CorrelationIDs, dlTime)), dlTime, xTimeUpper)
 				msg, err := dnmsg.MarshalJSON()
 				if err != nil {
 					logger.WithError(err).Error("Failed to marshal downlink message")
 					continue
 				}
 
-				logger.Info("Send downlink message")
+				logger.WithField("Downlink", dnmsg).Info("Send downlink message")
```
Comment (johanstokking) [inline pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: Remove this change. Note that this is `INFO` level

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 8]
```diff
@@ -129,6 +129,7 @@ func (s *srv) handleDiscover(c echo.Context) error {
 }
 
 func (s *srv) handleTraffic(c echo.Context) error {
+
```
Comment (johanstokking) [inline pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: Why?

**Comment by KrishnaIyer** [comment]
@johanstokking: I've fixed the minor typos and failing test.
For handing XTime, I think that it should be part of the connection since it's a "session" identifier.

If not, there are 2 possibilities that I see:
- put it in `conn.Context` and retrieve it for downlinks. I'm not in favor of this since it's not very visible.
- handle it locally in `handleTraffic` routine. This would work but it seems a bit ugly to me.

Please let me know your view.

**Pull request merged**